### PR TITLE
Breadcrumbs for servicing

### PR DESCRIPTION
**Pull request by schellap**
@gkhanna79 PTAL.

---
This change is [Reviewable](https://reviewable.io/reviews/dotnet/cli/2546)

### CODE REVIEW

File: src/corehost/common/pal.windows.cpp [hunk 1]
```diff
@@ -54,6 +55,17 @@ bool pal::find_coreclr(pal::string_t* recv)
...
 
+bool pal::touch_file(const pal::string_t& path)
+{
+    HANDLE hnd = ::CreateFileW(path.c_str(), 0, 0, NULL, CREATE_NEW, FILE_ATTRIBUTE_NORMAL, NULL);
+    if (hnd == INVALID_HANDLE_VALUE)
+    {
+        trace::verbose(_X("Failed to leave breadcrumb"), HRESULT_FROM_WIN32(GetLastError()));
+        return false;
+    }
+    return true;
```
Comment (schellap) [inline src/corehost/common/pal.windows.cpp]: CloseHandle

File: src/corehost/common/pal.unix.cpp [hunk 2]
```diff
@@ -112,11 +124,47 @@ bool pal::is_path_rooted(const pal::string_t& path)
...
+    recv->clear();
+    pal::string_t ext;
+    if (!get_default_extensions_directory(&ext))
+    {
+        return false;
+    }
+
+    append_path(&ext, _X("breadcrumbs"));
+    if (access(ext.c_str(), (R_OK | W_OK)) != 0)
+    {
```
Comment (schellap) [inline src/corehost/common/pal.unix.cpp]: Dir exists check

**Comment by schellap** [comment]
@gkhanna79 PTAL

Cc @ericstj @leecow

**Comment by schellap** [comment]
@ellismg for changes to `pal.unix.cpp`

**Comment by ericstj** [comment]
Can you share sample paths this creates?

**Comment by schellap** [comment]
Samples:

```
C:\ProgramData\Microsoft\NetFramework\BreadcrumbStore\netcore,System.AppContext

C:\ProgramData\Microsoft\NetFramework\BreadcrumbStore\netcore,System.AppContext,1.0.0
```

**Comment by ericstj** [comment]
Those are package IDs and versions: correct?

**Comment by schellap** [comment]
Right, `netcore,PackageId,PackageVersion`. I used comma as separator to match existing breadcrumbs and to not confuse with package or version namespacing.

File: src/corehost/common/pal.unix.cpp [hunk 3]
```diff
@@ -112,11 +124,64 @@ bool pal::is_path_rooted(const pal::string_t& path)
...
-    append_path(recv, _X("opt"));
-    append_path(recv, _X("dotnetextensions"));
+    pal::string_t ext;
+    if (pal::getenv(_X("DOTNET_EXTENSIONS"), &ext) && pal::realpath(&ext))
+    {
+        // We should have the path in ext.
+        trace::info(_X("Realpath $DOTNET_EXTENSIONS [%s]"), ext.c_str());
+    }
+    
+    if (!pal::directory_exists(ext))
```
Comment (gkhanna79) [inline src/corehost/common/pal.unix.cpp]: If DOTNET_EXTENSIONS is not defined, then will pal::string_t be zero initialized? If not, then this check is bogus. We should zero init ext and check it to be zero length here instead.

File: src/corehost/cli/breadcrumbs.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        m_breadcrumb_store.clear();
+    }
+}
+
+// Begin breadcrumb writing: write synchronously or launch a
+// thread to write breadcrumbs.
+void breadcrumb_writer_t::begin_write()
+{
+    trace::verbose(_X("--- Begin breadcrumb write"));
```
Comment (gkhanna79) [inline src/corehost/cli/breadcrumbs.cpp]: This trace should be moved to be after the empty check.

File: src/corehost/common/pal.unix.cpp (hunk 3 above)
Comment (schellap) [inline src/corehost/common/pal.unix.cpp]: It will be zero inited by the default constructor and directory exists will early out on an empty string.

File: src/corehost/cli/breadcrumbs.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    trace::verbose(_X("Number of breadcrumb files to write is %d"), m_files->size());
+    if (m_files->size() <= SYNCHRONOUS_WRITE_SIZE)
+    {
+        trace::verbose(_X("Writing breadcrumbs on the calling thread"));
+        m_status = write();
+        return;
+    }
+    m_thread = std::thread(thread_callback, this);
```
Comment (gkhanna79) [inline src/corehost/cli/breadcrumbs.cpp]: The callback should be named something like "breadcrumb_write_callback".

File: src/corehost/cli/breadcrumbs.cpp (hunk 4 above)
Comment (schellap) [inline src/corehost/cli/breadcrumbs.cpp]: No, this is simply tracking when the call was made.

File: src/corehost/cli/breadcrumbs.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    trace::verbose(_X("--- Begin breadcrumb write"));
+    if (m_breadcrumb_store.empty())
+    {
+        trace::verbose(_X("Breadcrumb store was not obtained... skipping write."));
+        m_status = false;
+        return;
+    }
+
+    trace::verbose(_X("Number of breadcrumb files to write is %d"), m_files->size());
+    if (m_files->size() <= SYNCHRONOUS_WRITE_SIZE)
```
Comment (gkhanna79) [inline src/corehost/cli/breadcrumbs.cpp]: How did we arrive at this? Do we have reason to believe that higher counts will be serviced better via another thread.

Comment (gkhanna79) [inline src/corehost/cli/breadcrumbs.cpp]: I would suggest doing all writes via a thread callback or none. If goal is to not block the main app thread, then move everything to thread callback.

File: src/corehost/cli/breadcrumbs.cpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        trace::warning(_X("An unexpected exception was thrown while leaving breadcrumbs"));
+    }
+    try
+    {
+        // Signal the waiting thread.
+        p_this->m_promise.set_value(retval);
+    }
+    catch (...)
+    {
+        trace::warning(_X("Could not signal waiting threads..."));
```
Comment (gkhanna79) [inline src/corehost/cli/breadcrumbs.cpp]: When can we be here?

File: src/corehost/cli/breadcrumbs.cpp (hunk 6 above)
Comment (schellap) [inline src/corehost/cli/breadcrumbs.cpp]: When CLI runs we emit 300 files i.e., roughly 300 probes and 300 writes in the worst case. We can always toggle the knob based on our perf studies. CLR leaves about 40 files at various points during assembly loads. Since we leave it eagerly I arrived at half that number. This is open to tuning.

File: src/corehost/cli/breadcrumbs.cpp [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        m_status = write();
+        return;
+    }
+    m_thread = std::thread(thread_callback, this);
+    trace::verbose(_X("Breadcrumbs will be written using a background thread"));
+}
+
+// Write the breadcrumbs. This method could be called fom the
+// calling thread and the background thread. But at any given
+// time only 1 thread will be here.
```
Comment (gkhanna79) [inline src/corehost/cli/breadcrumbs.cpp]: > But at any given time only 1 thread will be here.

There is no synchronization ensuring this. Is that expected or implied?

File: src/corehost/cli/breadcrumbs.cpp [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    for (const auto& file : *m_files)
+    {
+        pal::string_t file_path = m_breadcrumb_store;
+        pal::string_t file_name = _X("netcore,") + file;
+        append_path(&file_path, file_name.c_str());
+        if (!pal::file_exists(file_path))
+        {
+            if (!pal::touch_file(file_path))
+            {
+                successful = false;
```
Comment (gkhanna79) [inline src/corehost/cli/breadcrumbs.cpp]: @ericstj Is it reasonable to expect breadcrumb writing to fail?

Comment (gkhanna79) [inline src/corehost/cli/breadcrumbs.cpp]: What is the behavior on desktop for similar situation?

File: src/corehost/cli/breadcrumbs.cpp (hunk 8 above)
Comment (schellap) [inline src/corehost/cli/breadcrumbs.cpp]: It is implied since we launch a thread and forget about doing anything else in the main thread. I can add an assert we don't call begin write twice.

File: src/corehost/cli/fxr/hostfxr.cpp [hunk 10]
```diff
@@ -60,14 +60,18 @@ int execute_app(
 
     if (code != StatusCode::Success)
     {
-        trace::error(_X("Could not load host policy library from [%s]"), impl_dll_dir.c_str());
+        trace::error(_X("Expected to load %s from [%s]"), LIBHOSTPOLICY_NAME, impl_dll_dir.c_str());
```
Comment (gkhanna79) [inline src/corehost/cli/fxr/hostfxr.cpp]: I would suggest a more user-understandable message - "Invalid .NET Core FX configuration in folder <path to folder>"

File: src/corehost/common/pal.unix.cpp [hunk 11]
```diff
@@ -112,11 +124,64 @@ bool pal::is_path_rooted(const pal::string_t& path)
...
+    }
+
+    append_path(&ext, _X("breadcrumbs"));
+    if (!pal::directory_exists(ext))
+    {
+        trace::info(_X("Breadcrumb store does not exist in [%s]"), ext.c_str());
+        return false;
+    }
+
+    if (access(ext.c_str(), (R_OK | W_OK)) != 0)
```
Comment (ericstj) [inline src/corehost/common/pal.unix.cpp]: Is it possible that we can't enumerate perms but can actually write?

File: src/corehost/cli/breadcrumbs.cpp (hunk 5 above)
Comment (schellap) [inline src/corehost/cli/breadcrumbs.cpp]: It is already namespaced within `breadcrumb_writer_t`, so `breadcrumb_writer_t::breadcrumb_write_callback` is redundant. I've renamed it to `write_worker_callback`

File: src/corehost/cli/breadcrumbs.cpp (hunk 6 above)
Comment (schellap) [inline src/corehost/cli/breadcrumbs.cpp]: I like that suggestion. I've changed it to always launch a thread.

File: src/corehost/cli/breadcrumbs.cpp (hunk 7 above)
Comment (schellap) [inline src/corehost/cli/breadcrumbs.cpp]: This is my being extra careful, as an uncaught exception means a fail fast. Breadcrumb design philosophy is to not break the app in doing so.

Comment (schellap) [inline src/corehost/cli/breadcrumbs.cpp]: The downside is, the caller might block on the shared state which is in theory breaking as well. An exception at this point should not happen unless the app is already in an invalid state like a memory corruption or some unforeseen situation.

File: src/corehost/cli/fxr/hostfxr.cpp (hunk 10 above)
Comment (schellap) [inline src/corehost/cli/fxr/hostfxr.cpp]: Okay done.

File: src/corehost/common/pal.unix.cpp (hunk 11 above)
Comment (schellap) [inline src/corehost/common/pal.unix.cpp]: @ericstj, I haven't addressed your PR yet, as I wanted to test this out first. Will send another commit for this one specifically.

**Comment by schellap** [comment]
@dotnet-bot test Windows_NT x86 Debug Build please

**Comment by schellap** [comment]
@dotnet-bot test OSX x64 Release Build please

File: src/corehost/cli/fxr/hostfxr.cpp (hunk 10 above)
Comment (gkhanna79) [inline src/corehost/cli/fxr/hostfxr.cpp]: Is there a PR with this change?

Comment (gkhanna79) [inline src/corehost/cli/fxr/hostfxr.cpp]: Never mind - saw the change.

File: src/corehost/common/pal.unix.cpp [hunk 12]
```diff
@@ -112,11 +124,68 @@ bool pal::is_path_rooted(const pal::string_t& path)
...
+        // We should have the path in ext.
+        trace::info(_X("Realpath CORE_BREADCRUMBS [%s]"), ext.c_str());
+    }
+
+    if (!pal::directory_exists(ext))
+    {
+        trace::info(_X("Directory core breadcrumbs [%s] was not specified or found"), ext.c_str());
+        ext.clear();
+        append_path(&ext, _X("opt"));
+        append_path(&ext, _X("corebreadcrumbs"));
```
Comment (gkhanna79) [inline src/corehost/common/pal.unix.cpp]: Is it "breadcrumb" or "breadcrumbs" on Windows? We should ensure that whatever follows "CORE" is consistent across platforms.

Comment (schellap) [inline src/corehost/common/pal.unix.cpp]: In Windows it is called BreadcrumbStore. On Unix this is abbreviated partially as BreadcrumbS ;) On Windows there is no env variable. We are consistent on Unix on the plural for both env and dir.

**Comment by schellap** [comment]
@gkhanna79 the name core_servicing needs to be locked down for muxer lock down. We are okay on the branding etc.? @ericstj brought up a point that this is too generic as a name and I agree -- effectively, if we screw up inside this folder, customers won't know who to talk to.

**Comment by gkhanna79** [comment]
We do not expect the name to be known publicly - it is an infra-use artifact. And I expect that @leecow document on servicing to reference it as well.

**Comment by schellap** [comment]
Fixes #2607
